packages/wrangler/src/pipelines/client.ts

```diff
@@ -96,44 +102,84 @@ export type PermissionGroup = {
 	scopes: string[];
 };
 
+interface S3AccessKey {
+	accessKeyId: string;
+	secretAccessKey: string;
+}
+
 // Generate a Service Token to write to R2 for a pipeline
 export async function generateR2ServiceToken(
-	label: string,
 	accountId: string,
-	bucket: string
-): Promise<ServiceToken> {
-	const res = await fetchResult<PermissionGroup[]>(
-		`/user/tokens/permission_groups`,
-		{
-			method: "GET",
-		}
-	);
-	const perm = res.find(
-		(g) => g.name == "Workers R2 Storage Bucket Item Write"
-	);
-	if (!perm) {
-		throw new Error("Missing R2 Permissions");
-	}
-
-	// generate specific bucket write token for pipeline
-	const body = JSON.stringify({
-		policies: [
-			{
-				effect: "allow",
-				permission_groups: [{ id: perm.id }],
-				resources: {
-					[`com.cloudflare.edge.r2.bucket.${accountId}_default_${bucket}`]: "*",
-				},
-			},
-		],
-		name: label,
+	bucketName: string,
+	pipelineName: string
+): Promise<S3AccessKey> {
+	let server: http.Server;
+	let loginTimeoutHandle: ReturnType<typeof setTimeout>;
+	const timerPromise = new Promise<S3AccessKey>((_, reject) => {
+		loginTimeoutHandle = setTimeout(() => {
+			server.close();
+			clearTimeout(loginTimeoutHandle);
+			reject(
+				new UserError(
+					"Timed out waiting for authorization code, please try again."
+				)
+			);
+		}, 120000); // wait for 120 seconds for the user to authorize
 	});
 
-	return await fetchResult<ServiceToken>(`/user/tokens`, {
-		method: "POST",
-		headers: API_HEADERS,
-		body,
+	const loginPromise = new Promise<S3AccessKey>((resolve, reject) => {
```

Comment:
It was discussed with the IAM team, and we decided we do not want to do the PKCE flow as it opens up too many possibilities for impersonation. More details can be found here: https://wiki.cfdata.org/pages/viewpage.action?pageId=1053737124

Reply:
I was suggesting to extract the common parts, rather than to use the same flow exactly, but as I understand it you are under significant time pressure here. I can add a note to discuss refactoring and at a later date.

Reply:
Ah, gotcha. Yes that makes sense. Added a TODO for it. It would be great to have this pulled out, so we can test the handlers in each specific case.